pkg/engine/executor/pipeline_test.go

```diff
@@ -0,0 +1,494 @@
+		}
+
+		// Add each value to its respective builder
+		for i, value := range values {
+			value = strings.TrimSpace(value)
+			fieldBuilder := builder.Field(i)
+
+			// Skip nulls
+			if value == "" || value == "null" || value == "NULL" {
+				fieldBuilder.AppendNull()
+				continue
+			}
+
+			// Handle different types
+			switch fields[i].Type.ID() {
+			case arrow.INT64:
+				intVal, err := strconv.ParseInt(value, 10, 64)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse int64 value '%s': %w", value, err)
+				}
+				fieldBuilder.(*array.Int64Builder).Append(intVal)
+
+			case arrow.FLOAT64:
+				floatVal, err := strconv.ParseFloat(value, 64)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse float64 value '%s': %w", value, err)
+				}
+				fieldBuilder.(*array.Float64Builder).Append(floatVal)
+
+			case arrow.BOOL:
+				boolVal, err := strconv.ParseBool(value)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse bool value '%s': %w", value, err)
+				}
+				fieldBuilder.(*array.BooleanBuilder).Append(boolVal)
+
+			case arrow.INT32:
+				intVal, err := strconv.ParseInt(value, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse int32 value '%s': %w", value, err)
+				}
+				fieldBuilder.(*array.Int32Builder).Append(int32(intVal))
+
+			case arrow.FLOAT32:
+				floatVal, err := strconv.ParseFloat(value, 32)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse float32 value '%s': %w", value, err)
+				}
+				fieldBuilder.(*array.Float32Builder).Append(float32(floatVal))
+
+			default:
+				return nil, fmt.Errorf("unsupported field type for field '%s': %s", fields[i].Name, fields[i].Type)
+			}
+		}
+	}
+
+	// Build and return the record
+	record := builder.NewRecord()
+	return record, nil
+}
+
+// CSVPipeline implements the Pipeline interface and provides data from a CSV string.
+// It divides the data into batches of a specified size.
+type CSVPipeline struct {
+	fields []arrow.Field
+	schema *arrow.Schema
+
+	// Pre-built batches
+	batches []arrow.Record
+
+	// Current state
+	currentBatch int
+	currentErr   error
+}
+
+// NewCSVPipeline creates a new Pipeline from CSV data with a specified batch size.
+// It pre-builds all batches during initialization to ensure schema compatibility.
+func NewCSVPipeline(batchSize int, fields []arrow.Field, csvData string) (*CSVPipeline, error) {
+	if batchSize <= 0 {
+		return nil, fmt.Errorf("batch size must be positive, got %d", batchSize)
+	}
+
+	schema := arrow.NewSchema(fields, nil)
+
+	// Split CSV into lines
+	lines := strings.Split(strings.TrimSpace(csvData), "\n")
+	validLines := make([]string, 0, len(lines))
+	for _, line := range lines {
+		if line != "" {
+			validLines = append(validLines, line)
+		}
+	}
+
+	totalRows := len(validLines)
+	batchCount := (totalRows + batchSize - 1) / batchSize // Ceiling division
```

Comment:
Would it be easier to follow if the test passes multiple CSV records in so it controls when batching starts/ends? (It'd be more flexible at least, rather than having a fixed batch size)